src/crawler/crawler.ts

```diff
@@ -0,0 +1,261 @@
+import { CommandTypes, sendCommandToContent } from '@/bus/Command';
+
+interface CrawlerState {
+	isActive: boolean;
+	nextProcessTime: number;
+	rateLimit: number;
+}
+
+interface QueueUrlsResponse {
+	accepted: number;
+	rejected: number;
+	queueSize: number;
+	crawledCount: number;
+}
+
+interface NextUrlResponse {
+	url: string;
+}
+
+interface QueueUrlsRequest {
+	urls: string[];
+	sourceUrl: string;
+}
+
+class Crawler {
+	private readonly state: CrawlerState;
+	private process: ( html: string ) => Promise< void >;
+
+	constructor() {
+		this.state = {
+			isActive: false,
+			nextProcessTime: 0,
+			rateLimit: 1.0, // pages per sec; 1.0 means 1000ms delay between delays
+		};
+		// Initialize with empty process function
+		this.process = async () => {};
+	}
+
+	private log( level: 'log' | 'warn' | 'error', ...args: any[] ): void {
+		console[ level ]( ...args );
+	}
+
+	// Allow setting the process function
+	public setProcessFunction(
+		processFn: ( html: string ) => Promise< void >
+	): void {
+		this.process = processFn;
+	}
+
+	public async start(): Promise< void > {
+		if ( this.state.isActive ) {
+			this.log( 'log', 'Crawler already running' );
+			return;
+		}
+
+		this.state.isActive = true;
+		this.log( 'log', 'Crawler started' );
+
+		while ( this.state.isActive ) {
+			const next = await this.getNextUrl();
+			if ( next ) {
+				await this.processUrl( next );
+			} else {
+				this.state.isActive = false;
+				this.log( 'log', 'Crawler finished' );
+			}
+		}
+	}
+
+	private async processUrl( url: string ): Promise< void > {
+		this.log( 'log', 'processing url', url );
+		try {
+			// Wait until we're allowed to process the next URL
+			await this.waitForRateLimit();
+
+			await this.navigateToUrl( url );
+
+			// @TODO: Get the HTML content via bus?
+			const html = document.documentElement.outerHTML;
+
+			// Process the page content
+			await this.process( html );
+
+			// Extract and queue new URLs
+			const links = this.extractLinks( html );
+			await this.queueUrls( links, url );
+		} catch ( error ) {
+			this.log( 'error', 'Error processing URL', url, error );
+			this.state.isActive = false;
+		}
+	}
+
+	private async waitForRateLimit(): Promise< void > {
+		const now = Date.now();
+		const delayMs = 1000 / this.state.rateLimit; // Convert rate limit to milliseconds between requests
+
+		if ( now < this.state.nextProcessTime ) {
+			await new Promise( ( resolve ) =>
+				setTimeout( resolve, this.state.nextProcessTime - now )
+			);
+		}
+
+		// Calculate next allowed process time using the delay
+		this.state.nextProcessTime = now + delayMs;
+	}
+
+	private extractLinks( htmlString: string ): string[] {
```

Comment:
We have a [similar plumbing in PHP](https://github.com/WordPress/wordpress-playground/blob/571128bde1c2caf586d0462904952434ddaad30a/packages/playground/data-liberation/src/functions.php#L26):

```php
	$p = new WP_Block_Markup_Url_Processor( $options['block_markup'], $options['base_url'] );
	while ( $p->next_url() ) {
		$parsed_url = $p->get_parsed_url();
		foreach ( $url_mapping as $mapping ) {
			if ( url_matches( $parsed_url, $mapping['from_url'] ) ) {
				$p->replace_base_url( $mapping['to_url'] );
				break;
			}
		}
	}
```

See how it also matches the domains and paths to stay within the same site. It might be handy to delegate that work to PHP.